### PR TITLE
ompi/osc/ucx: remove global TLS tables

### PR DESCRIPTION
**Pull request by janjust**
Remove workerpool's global thread storage tables.

Co-authored with:
Artem Y. Polyakov artemp@mellanox.com

Signed-off-by: Tomislav Janjusic <tomislavj@mellanox.com>

### CODE REVIEW

**Comment by janjust** [comment]
bot:retest

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 1]
```diff
@@ -691,19 +690,13 @@ int ompi_osc_ucx_free(struct ompi_win_t *win) {
...
-      return ret;
-    }
+    opal_common_ucx_wpmem_free(module->state_mem);
+    opal_common_ucx_wpmem_free(module->mem);
 
     opal_common_ucx_wpctx_release(module->ctx);
 
-    if (module->disp_units) free(module->disp_units);
+    if (module->disp_units)
+        free(module->disp_units);
```
Comment (artpol84) [inline ompi/mca/osc/ucx/osc_ucx_component.c:699]:
```suggestion
    if (module->disp_units) {
        free(module->disp_units);
    }
```

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 2]
```diff
@@ -949,136 +687,83 @@ static int _tlocal_ctx_connect(_tlocal_ctx_t *ctx_rec, int target)
...
 
-    tls->mem_tbl[free_idx]->ctx_rec = ctx_rec;
-    OPAL_ATOMIC_ADD_FETCH32(&ctx_rec->refcnt, 1);
-
-    tls->mem_tbl[free_idx]->mem->worker = ctx_rec->winfo;
-    tls->mem_tbl[free_idx]->mem->rkeys = calloc(mem->ctx->comm_size,
-                                         sizeof(*tls->mem_tbl[free_idx]->mem->rkeys));
+    opal_mutex_lock(&mem->mutex);
+    opal_list_append(&mem->mem_records, &mem_rec->super);
+    opal_mutex_unlock(&mem->mutex);
```
Comment (artpol84) [inline opal/mca/common/ucx/common_ucx_wpool.c:738]: Must go after initialization. Or during the flush someone might try to access it before it is initialized
```suggestion
```

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 3]
```diff
@@ -949,136 +687,83 @@ static int _tlocal_ctx_connect(_tlocal_ctx_t *ctx_rec, int target)
...
-    /* Make sure that we completed all the data structures before
-     * placing the item to the list
-     * NOTE: essentially we don't need this as list append is an
-     * operation protected by mutex
-     */
+    mem_rec->gmem = mem;
+    mem_rec->ctx_rec = ctx_rec;
+    mem_rec->winfo = ctx_rec->winfo;
+    mem_rec->rkeys = calloc(mem->ctx->comm_size, sizeof(*mem_rec->rkeys));
+
```
Comment (artpol84) [inline opal/mca/common/ucx/common_ucx_wpool.c:745]:
```suggestion
    opal_mutex_lock(&mem->mutex);
    opal_list_append(&mem->mem_records, &mem_rec->super);
    opal_mutex_unlock(&mem->mutex);
```

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 4]
```diff
@@ -1186,37 +858,37 @@ opal_common_ucx_winfo_flush(opal_common_ucx_winfo_t *winfo, int target,
...
+    OPAL_LIST_FOREACH(ctx_rec, &ctx->ctx_records, _ctx_record_t) {
+        opal_common_ucx_winfo_t *winfo = ctx_rec->winfo;
         if ((scope == OPAL_COMMON_UCX_SCOPE_EP) &&
-                (NULL == item->ptr->endpoints[target])) {
+                (NULL == winfo->endpoints[target])) {
             continue;
         }
-        opal_mutex_lock(&item->ptr->mutex);
-        rc = opal_common_ucx_winfo_flush(item->ptr, target, OPAL_COMMON_UCX_FLUSH_B,
+        rc = opal_common_ucx_winfo_flush(winfo, target, OPAL_COMMON_UCX_FLUSH_B,
```
Comment (artpol84) [inline opal/mca/common/ucx/common_ucx_wpool.c:878]: does this function protect the worker with `winfo->mutex`?

File: opal/mca/common/ucx/common_ucx_wpool.h [hunk 5]
```diff
@@ -118,6 +120,7 @@ typedef struct {
  */
 typedef struct opal_common_ucx_winfo {
     opal_recursive_mutex_t mutex;
```
Comment (artpol84) [inline opal/mca/common/ucx/common_ucx_wpool.h:122]:
```suggestion
    opal_list_item_t super; //?
    opal_recursive_mutex_t mutex;
```

File: opal/mca/common/ucx/common_ucx_wpool.h [hunk 6]
```diff
@@ -196,34 +215,35 @@ opal_common_ucx_tlocal_fetch(opal_common_ucx_wpmem_t *mem, int target,
                                 ucp_ep_h *_ep, ucp_rkey_h *_rkey,
                                 opal_common_ucx_winfo_t **_winfo)
 {
-    opal_common_ucx_tlocal_fast_ptrs_t *fp = NULL;
+    _mem_record_t *mem_rec = NULL;
     int expr;
```
Comment (artpol84) [inline opal/mca/common/ucx/common_ucx_wpool.h:219]:
```suggestion
    int is_ready;
```

File: opal/mca/common/ucx/common_ucx_wpool.h [hunk 7]
```diff
@@ -196,34 +215,35 @@ opal_common_ucx_tlocal_fetch(opal_common_ucx_wpmem_t *mem, int target,
...
-    rc = opal_tsd_tracked_key_get(&mem->mem_tls_key, (void**)&fp);
+    rc = opal_tsd_tracked_key_get(&mem->tls_key, (void**)&mem_rec);
     if (OPAL_SUCCESS != rc) {
         return rc;
     }
-    expr = fp && (NULL != fp->winfo) && (fp->winfo->endpoints[target]) &&
-            (NULL != fp->rkeys[target]);
+    expr = mem_rec && (NULL != mem_rec->winfo) && (mem_rec->winfo->endpoints[target]) &&
+            (NULL != mem_rec->rkeys[target]);
     if (OPAL_UNLIKELY(!expr)) {
```
Comment (artpol84) [inline opal/mca/common/ucx/common_ucx_wpool.h:229]:
```suggestion
    if (OPAL_UNLIKELY(!is_ready)) {
```

File: opal/mca/common/ucx/common_ucx_wpool.h [hunk 8]
```diff
@@ -196,34 +215,35 @@ opal_common_ucx_tlocal_fetch(opal_common_ucx_wpmem_t *mem, int target,
...
     /* First check the fast-path */
-    rc = opal_tsd_tracked_key_get(&mem->mem_tls_key, (void**)&fp);
+    rc = opal_tsd_tracked_key_get(&mem->tls_key, (void**)&mem_rec);
     if (OPAL_SUCCESS != rc) {
         return rc;
     }
-    expr = fp && (NULL != fp->winfo) && (fp->winfo->endpoints[target]) &&
-            (NULL != fp->rkeys[target]);
+    expr = mem_rec && (NULL != mem_rec->winfo) && (mem_rec->winfo->endpoints[target]) &&
+            (NULL != mem_rec->rkeys[target]);
```
Comment (artpol84) [inline opal/mca/common/ucx/common_ucx_wpool.h:228]: Is it possible that mem_rec exists, but mem_rec->winfo is not set?
maybe it should go to become an assert
```suggestion
    is_ready = mem_rec && (mem_rec->winfo->endpoints[target]) &&
            (NULL != mem_rec->rkeys[target]);
    MCA_COMMON_UCX_ASSERT((NULL == mem_rec) || (NULL != mem_rec->winfo));
```

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 9]
```diff
@@ -146,13 +153,12 @@ opal_common_ucx_wpool_init(opal_common_ucx_wpool_t *wpool,
...
     }
 
-    OBJ_CONSTRUCT(&wpool->mutex, opal_recursive_mutex_t);
-    OBJ_CONSTRUCT(&wpool->tls_list, opal_list_t);
+    OBJ_CONSTRUCT(&wpool->mutex, opal_mutex_t);
 
     status = ucp_config_read("MPI", NULL, &config);
     if (UCS_OK != status) {
         MCA_COMMON_UCX_VERBOSE(1, "ucp_config_read failed: %d", status);
-        return OPAL_ERROR;
```
[the remark this answers is not in the document]
Comment (janjust) [reply opal/mca/common/ucx/common_ucx_wpool.c:155]: indeed, fixed

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 10]
```diff
@@ -421,26 +415,23 @@ opal_common_ucx_wpctx_create(opal_common_ucx_wpool_t *wpool, int comm_size,
...
 
     /* Make sure that all the loads/stores are complete */
     opal_atomic_mb();
 
-    /* If there is no more references to this handler
-     * we can release it */
-    if (0 == my_refcntr) {
-        _common_ucx_wpctx_free(ctx);
-    }
+    _common_ucx_wpctx_free(ctx);
```
[the remark this answers is not in the document]
Comment (janjust) [reply opal/mca/common/ucx/common_ucx_wpool.c:433]: yes, the _common_ucx_wpctx_free() was a small function with a single use, I didn't even inline it, just copied the code to the caller's function. I think it's more intuitive that way.

File: opal/mca/common/ucx/common_ucx_wpool.c (hunk 4 above)
Comment (janjust) [reply opal/mca/common/ucx/common_ucx_wpool.c:877]: It does not - I moved the winfo->mutex to include the winfo_flush

File: opal/mca/common/ucx/common_ucx_wpool.h (hunk 5 above)
Comment (janjust) [reply opal/mca/common/ucx/common_ucx_wpool.h:122]: Why are we adding this? We're not storing the winfo as an opal list item. but instead we have a seperate winfo list item which includes the pointer to the winfo.

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 11]
```diff
@@ -955,130 +705,69 @@ static int _tlocal_ctx_connect(_tlocal_ctx_t *ctx_rec, int target)
...
+    _mem_record_t *mem_rec = (_mem_record_t *)arg;
+    if (NULL == mem_rec) {
+        return;
+    }
 
     for(i = 0; i < mem_rec->gmem->ctx->comm_size; i++) {
-        if (mem_rec->mem->rkeys[i]) {
-            ucp_rkey_destroy(mem_rec->mem->rkeys[i]);
+        if (mem_rec->rkeys[i]) {
+            ucp_rkey_destroy(mem_rec->rkeys[i]);
```
[the remark this answers is not in the document]
Comment (janjust) [reply opal/mca/common/ucx/common_ucx_wpool.c:712]: This is called from a thread destructor, thus it's not a critical section - the critical section is adding/removing mem_records from the list - which is locked with a wpmem->mutex

File: opal/util/net.c [hunk 12]
```diff
@@ -474,6 +474,11 @@ opal_net_init()
 int
 opal_net_finalize()
 {
+#if OPAL_ENABLE_IPV6
```
[the remark this answers is not in the document]
Comment (janjust) [reply opal/util/net.c:477]: You're right - hmm, not sure what happened here...

[the remark this answers is not in the document]
Comment (janjust) [reply opal/util/net.c:477]: Probably a rebase gone awry

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 13]
```diff
@@ -651,287 +560,116 @@ static int _comm_ucx_wpmem_map(opal_common_ucx_wpool_t *wpool,
...
+    opal_list_prepend(&wpool->idle_workers, &winfo->self->super);
+    opal_mutex_lock(&ctx_rec->gctx->mutex);
+    opal_list_remove_item(&ctx_rec->gctx->ctx_records, &ctx_rec->super);
+    opal_mutex_unlock(&ctx_rec->gctx->mutex);
 
-    /* Remove myself from the communication context structure
-     * This may result in context release as we are using
-     * delayed cleanup */
-    _common_ucx_wpctx_remove(ctx_rec->gctx, ctx_rec->winfo);
+    opal_mutex_unlock(&wpool->mutex);
```
[the remark this answers is not in the document]
Comment (janjust) [reply opal/mca/common/ucx/common_ucx_wpool.c:625]: Why are we removing this unlock? We're modifying the wpool active/idle workers here.

File: opal/mca/common/ucx/common_ucx_wpool.c [hunk 14]
```diff
@@ -651,287 +560,116 @@ static int _comm_ucx_wpmem_map(opal_common_ucx_wpool_t *wpool,
...
-        //TODO: error out
+    /* Add ctx_rec to list */
+    opal_mutex_lock(&ctx->mutex);
+    opal_list_append(&ctx->ctx_records, &ctx_rec->super);
+    opal_mutex_unlock(&ctx->mutex);
+
+    /* Add tls reference to record */
+    rc = opal_tsd_tracked_key_set(&ctx->tls_key, ctx_rec);
+    if (OPAL_SUCCESS != rc) {
+        OBJ_RELEASE(ctx_rec);
```
[the remark this answers is not in the document]
Comment (janjust) [reply opal/mca/common/ucx/common_ucx_wpool.c:663]: I don't agree, if we failed at any point, we're basically out of resources, and in the caller's functions we return a OPALL_ERR_OUT_OF_RESOURCE, why isn't this enough?

File: opal/mca/common/ucx/common_ucx_wpool.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (artpol84) [reply opal/mca/common/ucx/common_ucx_wpool.c:706]: You need to hold winfo mutex when you access worker

**Pull request merged**